support/scheduler/schedule.go

```diff
@@ -0,0 +1,442 @@
+			context := ScheduleContext{
+				ScheduleEventsMap: make(map[string]models.ScheduleEvent),
+				MarkedDeleted:     false,
+			}
+			context.Reset(schedule)
+
+			addScheduleOperation(newScheduleId, &context)
+		}
+
+		addScheduleEventOperation(newScheduleId, scheduleEventId, scheduleEvent)
+	} else { // if not, just update the schedule event in place
+		scheduleContext.ScheduleEventsMap[scheduleEventId] = scheduleEvent
+	}
+
+	loggingClient.Info("updated the schedule event with id " + scheduleEvent.Id.Hex() + " to schedule id : " + schedule.Id.Hex())
+
+	return nil
+}
+
+func removeScheduleEvent(scheduleEventId string) error {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	loggingClient.Info("removing the schedule event with id " + scheduleEventId)
+
+	scheduleId, exists := scheduleEventIdToScheduleIdMap[scheduleEventId]
+	if !exists {
+		loggingClient.Error("can not find schedule id with schedule event id : " + scheduleEventId)
+		return errors.New("can not find schedule id with schedule event id : " + scheduleEventId)
+	}
+
+	scheduleContext, exists := scheduleIdToContextMap[scheduleId]
+	if !exists {
+		loggingClient.Error("can not find schedule context with schedule id : " + scheduleId)
+		return errors.New("can not find schedule context with schedule id : " + scheduleId)
+	}
+
+	delete(scheduleContext.ScheduleEventsMap, scheduleEventId)
+
+	//if there are no more events for the schedule, remove the schedule context
+	if len(scheduleContext.ScheduleEventsMap) == 0 {
+		loggingClient.Info("there are no more events for the schedule : " + scheduleId + ", remove it.")
+		deleteScheduleOperation(scheduleId, scheduleContext)
+	}
+
+	loggingClient.Info("removed the schedule event with id " + scheduleEventId)
+
+	return nil
+}
+
+//endregion
+
+//region schedule event execution
+func triggerSchedule() {
+	nowEpoch := time.Now().Unix()
+
+	defer func() {
+		if err := recover(); err != nil {
+			loggingClient.Error("trigger schedule error : " + err.(string))
+		}
+	}()
+
+	if scheduleQueue.Length() == 0 {
+		return
+	}
+
+	loggingClient.Debug(fmt.Sprintf("%d item in schedule queue.", scheduleQueue.Length()))
+
+	for i := 0; i < scheduleQueue.Length(); i++ {
+		if scheduleQueue.Peek().(*ScheduleContext) != nil {
+			scheduleContext := scheduleQueue.Remove().(*ScheduleContext)
+			scheduleId := scheduleContext.Schedule.Id.Hex()
+			loggingClient.Debug("check schedule with id : " + scheduleId)
+			if scheduleContext.MarkedDeleted {
+				loggingClient.Debug("the schedule with id : " + scheduleId + " be marked as deleted, removing it.")
+				continue //really delete from the queue
+			} else {
+				loggingClient.Debug("schedule with id : " + scheduleId + " next schedule time is : " + scheduleContext.NextTime.String())
+				if scheduleContext.NextTime.Unix() <= nowEpoch {
+					loggingClient.Info("executing schedule, detail : {" + scheduleContext.GetInfo() + "} , at : " + scheduleContext.NextTime.String())
+					execute(scheduleContext)
```

Comment:
Would it be worthwhile to decouple the execution of a given schedule from the timer handling? In this case, we're looping through schedules one at a time. If one is ready to fire we run it. Could we not utilize a channel, publishing a message to that channel where an async handler takes over?

You could also probably refactor StartTicker with a channel whereby the consumer continuously reads, comparing the current time to the triggering time of the schedule. If there isn't a match, just put it back into the queue. That would eliminate the need for the ticker.

A channel IS a queue, so you wouldn't need custom logic or types to manage that.

Reply:
this implementation I refer to the old code, but your thinking is great. I will try to refactor my code based on your idea, give me some time, thanks.

Reply:
@tsconn23 I considered replaced the specific queue with a channel. But the channel's buffer num is fixed. If the buffer is full the producer will be blocked so we should resize. The queue package which I used has the `resize` function. Actually, a channel's like **queue** + **sync(lock)** , and need resize manually. The current implementation is **queue(array with resize function)** + **sync**. So is there necessary to change the current implementation. I am considering, what's your opinion?

Reply:
@yanghua You could leave the queue as is for now, but I would still consider invoking execute(scheduleContext) on a separate goroutine. One unknown w/r/t the queue refactoring is the volume of schedules to be handled. We can wait and see.